react/src/components/Chat/ChatMessageContent.tsx

```diff
@@ -1,89 +1,143 @@
+import Flex from '../Flex';
 import CopyButton from './CopyButton';
-import { Card, Typography } from 'antd';
+import { SyntaxHighlighter, SyntaxHighlighterProps } from './SyntaxHighight';
+import { Typography } from 'antd';
 // `rehype-katex` does not import the CSS file, so we need to import it manually.
 import 'katex/dist/katex.min.css';
-import React from 'react';
+import React, { memo, useCallback } from 'react';
 import Markdown from 'react-markdown';
-import {
-  Prism as SyntaxHighlighter,
-  SyntaxHighlighterProps,
-} from 'react-syntax-highlighter';
 import { oneLight } from 'react-syntax-highlighter/dist/esm/styles/prism';
 import rehypeKatex from 'rehype-katex';
 import remarkGfm from 'remark-gfm';
 import remarkMath from 'remark-math';
 
 const { Text } = Typography;
 
+const LangType = memo<{ lang: string }>(({ lang }) => {
+  return (
+    <Text style={{ fontWeight: 'normal' }} type="secondary">
+      {lang}
+    </Text>
+  );
+});
+
+LangType.displayName = 'RenderLangType';
+
+const CodeBlock = memo<SyntaxHighlighterProps>(({ children, ...props }) => (
+  <SyntaxHighlighter {...props}>{children}</SyntaxHighlighter>
+));
+
+CodeBlock.displayName = 'CodeBlock';
+
+const CodeHead = memo<{ lang: string; extra?: React.ReactNode }>(
+  ({ lang, extra }) => (
+    <Flex
+      style={{
+        margin: '0',
+        minHeight: '38px',
+        padding: '0 12px',
+        background: 'rgba(0, 0, 0, 0.02)',
+        width: '100%',
+      }}
+    >
+      <Flex
+        style={{
+          display: 'inline-block',
+          flex: '1',
+          overflow: 'hidden',
+          whiteSpace: 'nowrap',
+          textOverflow: 'ellipsis',
+        }}
+      >
+        <LangType lang={lang} />
+      </Flex>
+      <Flex>{extra}</Flex>
+    </Flex>
+  ),
+  (prevProps, nextProps) => {
+    if (prevProps.lang !== nextProps.lang) return false;
+    if (prevProps.extra !== nextProps.extra) return false;
+    return true;
+  },
+);
+
+CodeHead.displayName = 'CodeHead';
+
 const ChatMessageContent: React.FC<{
   children?: string;
-}> = ({ children }) => {
+  isStreaming?: boolean;
+}> = ({ children, isStreaming }) => {
+  const renderCode = useCallback(
```

Comment:
using useCallback to preventing re-render code block which were already rendered before. It makes improving render performance on syntax highlighting